### PR TITLE
[MM-56073] MMCTL delete post command

### PR DESCRIPTION
**Pull request by BenCookie95**
#### Summary
Updated the delete post endpoint to take a new url parameter called `permanent` that allows for a post and it's attachments to be hard deleted from the db, file store and search indexer. Only system admins can make this request.

Added a new config called `ServiceSettings.EnableAPIPostDeletion` in order to enable/disable post deletion. This config does not need to be enabled when running mmctl in local mode.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-56073

#### Release Note

```release-note
- Added a new URL parameter called permanent to DELETE /api/v4/posts/<post-id>. Set permanent to true in order to permanently delete a post and it's attachments.
- Added a new mmctl command, mmctl post delete <post-id>, in order to permanently delete a post and it's attachments.
- Added a new config called ServiceSettings.EnableAPIPostDeletion in order to enable/disable post deletion. This config does not need to be enabled when running mmctl in local mode.
```

### CODE REVIEW

**Comment by hanzei** [comment]
@BenCookie95 Heads up that I merged `master` as that hopefully fixes CI.

File: server/channels/app/post.go
Comment (BenCookie95) [inline server/channels/app/post.go:2643]: We publish a few websockets in different places with the `post.message` contents after deletion, I initially thought this was an issue because posts are sometimes deleted due to security concerns of the message contents. I tried just setting `post.message` to `""` here but `RemoveNotifications` needs the message contents in order to figure out if there are mentions in the message. I was thinking of just removing the message contents for `WebsocketEventPostDeleted` websockets but we can just leave it in for the `WebsocketEventThreadUpdated` websocket. The `WebsocketEventThreadUpdated` websocket only gets published if a thread comment is deleted

Comment (BenCookie95) [reply server/channels/app/post.go:2643]: It could also be a breaking change to `hooks.MessageHasBeenDeleted` if we remove the message contents

**Review by pvev — APPROVED** [review]
LGTM! thanks @BenCookie95 🚀 , just some questions and some nits.

File: webapp/channels/src/packages/mattermost-redux/src/reducers/entities/posts.ts
Comment (hanzei) [inline webapp/channels/src/packages/mattermost-redux/src/reducers/entities/posts.ts:197]: These changes seem unrelated. Are they needed for the PR?

File: server/cmd/mmctl/commands/post.go
Comment (hanzei) [inline server/cmd/mmctl/commands/post.go:49]: I wonder if we should use a "real" post ID to demonstrate what kind of data the user should provide here.

File: server/cmd/mmctl/commands/post.go
Comment (hanzei) [inline server/cmd/mmctl/commands/post.go:68]: @BenCookie95 We did talk offline about adding a `--permanent` flag that enables the hard deletion and, by default, only soft deleting the post. Is that still a proposal that you want to implement?

File: server/cmd/mmctl/commands/post_test.go
Comment (hanzei) [inline server/cmd/mmctl/commands/post_test.go:1]: Would you please implement a simple e2e test for confidence?

File: server/cmd/mmctl/commands/post.go
Comment (hanzei) [inline server/cmd/mmctl/commands/post.go:253]: Could this method return an error so mmctl returns a non-zero error code, as in https://github.com/mattermost/mattermost/pull/26428?

File: server/channels/api4/post.go
Comment (hanzei) [inline server/channels/api4/post.go:672]: 2/5 that this should be a 501. We use the error code in other parts of the code when a config setting is missing, e.g. in https://github.com/mattermost/mattermost/blob/aa85a13c8f0d6a5cfd24f90e7b529076055d2638/server/channels/api4/drafts.go#L73-L77.

File: server/channels/api4/post.go
Comment (hanzei) [inline server/channels/api4/post.go:654]: Do we need to be that restrictive? I expect that users can delete their own posts by default. That would be very helpful to have for GDPR compliance on public servers.

Should we introduce a new permission?

File: server/channels/api4/post.go
Comment (hanzei) [inline server/channels/api4/post.go:669]: 3/5 that this check should happen before we load the post from the DB.

File: server/channels/api4/post_local.go
Comment (hanzei) [inline server/channels/api4/post_local.go:20]: Can we combine this function and `deletePost` to keep the code DRY?

File: server/channels/api4/post_test.go
Comment (hanzei) [inline server/channels/api4/post_test.go:2818]: Sub tests :heart:

File: server/public/model/config.go
Comment (hanzei) [inline server/public/model/config.go:385]: It's fine for consistency, but I still wonder why we need these settings at all. Should they be permission settings?

**Review by hanzei — CHANGES_REQUESTED** [review]
This will be a quite helpful command :rocket: 

A few high level comments below

File: server/public/model/config.go
Comment (BenCookie95) [reply server/public/model/config.go:385]: I think it makes sense to have these as configs and separate from the permissions. With the permanent nature of these deletions I think the extra safe guard makes sense

File: server/channels/api4/post_local.go
Comment (BenCookie95) [reply server/channels/api4/post_local.go:20]: They're a bit different due to the exclusion of the permissions in the local function, not much gained by combining the other parts of the function

File: server/channels/api4/post.go
Comment (BenCookie95) [reply server/channels/api4/post.go:654]: I put it at the same level as the other permanent deletion tasks because the ask was for sysadmins to do this and for other compliance reasons I think it's better to just do it this way for now. Regular users can use the soft delete and if they want a GDPR deletion they can talk to the sysadmin

File: webapp/channels/src/packages/mattermost-redux/src/reducers/entities/posts.ts
Comment (BenCookie95) [reply webapp/channels/src/packages/mattermost-redux/src/reducers/entities/posts.ts:197]: I didn't think there was a need to persist the message in redux state so I just removed it

File: server/channels/api4/post_local.go
Comment (BenCookie95) [reply server/channels/api4/post_local.go:46]: We already call `defer c.LogAuditRecWithLevel(auditRec, app.LevelContent)` further up and the app layer will return an app error already. So I think we're good?

File: server/channels/app/file.go
Comment (BenCookie95) [reply server/channels/app/file.go:1621]: We will have to loop over each fileinfo record at some stage in order to execute the deletion steps. I could potentially move the below calls to `RemoveFileFromFileStore` into go routines but no need right now I think.
```
a.RemoveFileFromFileStore(c, info.Path)
if info.PreviewPath != "" {
	a.RemoveFileFromFileStore(c, info.PreviewPath)
}
if info.ThumbnailPath != "" {
	a.RemoveFileFromFileStore(c, info.ThumbnailPath)
}
```

File: server/channels/app/file.go
Comment (BenCookie95) [reply server/channels/app/file.go]: I ended up returning a not found error so that it's easier to ignore from the caller if you want to.

File: server/channels/api4/post.go
Comment (BenCookie95) [reply server/channels/api4/post.go]: Moved this up above the GetPost call along with the session check

**Comment by BenCookie95** [comment]
@hanzei Addressed most concerns, I just need to push the e2e test tomorrow

File: server/cmd/mmctl/commands/post.go
Comment (hanzei) [inline server/cmd/mmctl/commands/post.go:55]: Can we simplify that?
```suggestion
  $ mmctl post delete udjmt396tjghi8wnsk3a1qs1sw --permanent

  # Permanently delete multiple posts and their file contents from the database and filestore
  $ mmctl post delete udjmt396tjghi8wnsk3a1qs1sw 7jgcjt7tyjyyu83qz81wo84w6o --permanent`,
```

File: server/cmd/mmctl/commands/post.go
Comment (hanzei) [inline server/cmd/mmctl/commands/post.go:271]: Do we update `result`?

File: server/cmd/mmctl/commands/post.go
Comment (hanzei) [inline server/cmd/mmctl/commands/post.go:47]:
```suggestion
	Long:  `This command will mark the post as deleted and remove it from the user's clients, but it does not permanently delete the post from the database. Please use the --permanent flag to permanently delete a post and its attachments from your database.`,
```

File: server/channels/api4/post.go
Comment (hanzei) [reply server/channels/api4/post.go]: Fair point. I hoped to squeeze in some GDPR-related changes into the PR, but consistency is important.

File: server/channels/app/file.go
Comment (pvev) [reply server/channels/app/file.go:1621]: yes, that was more like what I was referring to, move one layer up but you are right, no needed now

File: server/channels/store/searchlayer/file_info_layer.go
Comment (hanzei) [inline server/channels/store/searchlayer/file_info_layer.go:131]: Did `deleteFileIndexForPost` ever work? I don't understand if the change is needed because this PR changes things or if `deleteFileIndexForPost` has been broken before.

**Review by hanzei — CHANGES_REQUESTED** [review]
I'm a bit confused by the search layer changes. Maybe you can help me understand the reason behind the changes in `server/channels/store/searchlayer/file_info_layer.go`.

File: server/channels/store/searchlayer/file_info_layer.go
Comment (BenCookie95) [reply server/channels/store/searchlayer/file_info_layer.go:131]: I don't think deleteFileIndexForPost has ever worked. I filed this ticket for it https://mattermost.atlassian.net/browse/MM-59264

File: server/channels/app/post.go
Comment (BenCookie95) [reply server/channels/app/post.go:2667]: We kept the same client websocket action for soft deletion vs hard deletion. Do we just leave the plugin hooks as is for now? I don't think this is a breaking change because we are giving them the post struct same as before and our apis will not return these deleted posts in both cases

File: server/channels/app/file.go
Comment (BenCookie95) [reply server/channels/app/file.go:1615]: I think that's in order to avoid a race condition due to `App.DeletePost` using a go routine for `deletePostFiles`

File: server/channels/app/file_test.go
Comment (BenCookie95) [reply server/channels/app/file_test.go:790]: Removed that error

File: server/channels/app/file.go
Comment (BenCookie95) [reply server/channels/app/file.go:1632]: I think having the logging here makes sense, I don't think there will be a caller who won't want it

File: server/channels/app/post.go
Comment (BenCookie95) [reply server/channels/app/post.go]: Removed that case, returning nil instead since it is a no-op

File: server/channels/app/post.go
Comment (hanzei) [reply server/channels/app/post.go:2667]: The documentation of `MessageHasBeenDeleted` already makes it clear that the hook is invoked after the post was deleted from the DB (which was not true). This PR just adjusts the behaviour to match the docs. I think we are good to continue with your proposed changes. 

https://github.com/mattermost/mattermost/blob/b1fd445fcc197d2b5ad87ae0f21bdbfadebce4bf/server/public/plugin/hooks.go#L190-L195

File: server/cmd/mmctl/commands/post_test.go
Comment (hanzei) [inline server/cmd/mmctl/commands/post_test.go:1]: It's great that you added an E2E test using Cypress. That is great test coverage! :tada: 

Actually, I was talking about a go E2E test similar to the ones in `post_e2e_test.go`.I would prefer to see a basic test being added to `post_e2e_test.go`, but it's surely not a blocker.

File: server/channels/app/file.go
Comment (hanzei) [reply server/channels/app/file.go:1615]: If I understand the code correctly, there can be two entries in the `FileInfoCache` for the same post, one indexed with the `PostID` and the other indexed with `PostID_deleted`. The latter is used when the post is marked as deleted

I _think_ we are fine with not calling `InvalidateFileInfosForPostCache` as `DeletePost` should have cleared the cache entry index with `PostID`. But I wonder if we should still clear both cache entries just to be on the safe side. (And add explain the concerns via a comment)

**Review by hanzei — APPROVED** [review]
Great work, Ben :tada:

File: server/channels/app/file.go
Comment (BenCookie95) [reply server/channels/app/file.go:1615]: I took a look through our code and you're right, I think we need both. I'm confused why we do this with our cache keys because it doesn't make much sense but looks at this code snippet:
```go
func (s LocalCacheFileInfoStore) GetForPost(postId string, readFromMaster, includeDeleted, allowFromCache bool) ([]*model.FileInfo, error) {
	if !allowFromCache {
		return s.FileInfoStore.GetForPost(postId, readFromMaster, includeDeleted, allowFromCache)
	}

	cacheKey := postId
	if includeDeleted {
		cacheKey += "_deleted"
	}

	var fileInfo []*model.FileInfo
	if err := s.rootStore.doStandardReadCache(s.rootStore.fileInfoCache, cacheKey, &fileInfo); err == nil {
		return fileInfo, nil
	}

	fileInfos, err := s.FileInfoStore.GetForPost(postId, readFromMaster, includeDeleted, allowFromCache)
	if err != nil {
		return nil, err
	}

	if len(fileInfos) > 0 {
		s.rootStore.doStandardAddToCache(s.rootStore.fileInfoCache, cacheKey, fileInfos)
	}

	return fileInfos, nil
}
```
If someone includes the `include_deleted` url parameter it will append `_deleted` to the cache key, it won't find it in the cache layer and then it will query the db. The db will return the file, doesn't matter if the file is deleted or not because `include_deleted` means we want it either way. We then create a cache key with `_deleted` appended without actually checking if the file is deleted. If someone tries to fetch that file without the `include_deleted` query param it will skip the cache layer and query the db. We probably need to file a ticket to fix this and only append `_deleted` for files that are marked as deleted. I'm not even sure why we append `_deleted` tbh

**Comment by DHaussermann** [comment]
Hi @BenCookie95 sorry for the delay on this. I have done a round of testing and see both the hard delete and soft delete working as expected. Also hard deleted a post that was already soft deleted. 

I do have a couple very minor points to raise. 
1. This is maybe trivial but the `delete` subcommand does not autocomplete the way the other to `post` options do. When I type  `mmctl post del` and then hit `tab` there is no completion found. This may cause confusion.  
2. The response if I attempt to soft deleted a post that is already soft deleted is not very graceful 
image
Ideally we could respond with feedback about the operation being invalid instead of this. Is there a low effort way to check this case?

**Comment by hanzei** [comment]
1. @DHaussermann  Did you reload reload your shell after you installed the `mmctl` version from this branch?
2. @BenCookie95 The permission checks for the local endpoint seem wrong. Can you please double-check the code?

**Comment by hanzei** [comment]
@BenCookie95 Heads up that there is a conflict to resolve

**Comment by DHaussermann** [comment]
> 1. @DHaussermann  Did you reload reload your shell after you installed the `mmctl` version from this branch?
> 2. @BenCookie95 The permission checks for the local endpoint seem wrong. Can you please double-check the code?

For 1. I have confirmed Ben was correct. the autocomplete **does** exist I just had to reload the shell.

**Comment by BenCookie95** [comment]
> @BenCookie95 The permission checks for the local endpoint seem wrong. Can you please double-check the code?

The local endpoints do not require permission checks as they will be bypassed. The local socket does not have a session

**Comment by BenCookie95** [comment]
@enzowritescode Due to the nature of this change we thought it should go through a security review before being merged. Let me know if you have any questions

**Review by enzowritescode — CHANGES_REQUESTED** [review]
Overall looks good to me! Some very minor feedback

File: server/public/model/client4.go
Comment (hanzei) [reply server/public/model/client4.go]: Fixed

File: server/channels/store/timerlayer/timerlayer.go
Comment (hanzei) [reply server/channels/store/timerlayer/timerlayer.go]: Good point. I updated the code to be confident with the rest of the code base.

File: server/channels/api4/post.go
Comment (enzowritescode) [inline server/channels/api4/post.go:636]: Just so I understand, why the separate variable that only gets used in one spot?

File: server/channels/api4/post_local.go
Comment (enzowritescode) [inline server/channels/api4/post_local.go:32]: Same question as before. Curious the significance of using this variable when `permanent` is used in others?

**Review by enzowritescode — COMMENTED** [review]
Two quick questions

File: server/channels/api4/post.go
Comment (hanzei) [reply server/channels/api4/post.go:636]: `includeDeleted` is the argument name for `GetSinglePost`. 0/5 that it's easier to read this way, as the call to `GetSinglePost` is easier to understand.

**Review by enzowritescode — APPROVED** [review]

**Comment by hanzei** [comment]
@DHaussermann The points you raised have been addressed. PTAL.

**Comment by DHaussermann** [comment]
Sorry @BenCookie95 I can't build this now. Are you able to solve the conflicts?

**Comment by BenCookie95** [comment]
> Sorry @BenCookie95 I can't build this now. Are you able to solve the conflicts?

@DHaussermann Conflicts resolved, hopefully pipeline passes

**Comment by DHaussermann** [comment]
Hi @BenCookie95,
The issue above is resolved. One last thing I noticed here is that `EnableAPIPostDeletion` is enforced when you use an authenticated user but, when you use `--local` it does not seem to check this. 
image

I'm not that familiar with MMCTL can you confirm if this is expected?

**Comment by BenCookie95** [comment]
> Hi @BenCookie95, The issue above is resolved. One last thing I noticed here is that `EnableAPIPostDeletion` is enforced when you use an authenticated user but, when you use `--local` it does not seem to check this. image
> 
> I'm not that familiar with MMCTL can you confirm if this is expected?

@DHaussermann That is expected for local mode, we do the same for permanent user deletion and permanent team deletion

**Review by DHaussermann — APPROVED** [review]
Tested and passed 
The new delete post command is functioning correctly based on testing above.

LGTM!

**Comment by oscrx** [comment]
This doesn't seem to be documented yet or am I mistaken?
https://api.mattermost.com/#tag/posts/operation/DeletePost

**Comment by hanzei** [comment]
Good catch @oscrx :+1: 

@BenCookie95 Would you mind updating the API docs?

**Pull request merged**
